### PR TITLE
feat(application-components): add experimental `product-picker-input`

### PR DESCRIPTION
**Pull request by adnasa**
#### Summary

- adds `debounce`, `use-load-with-prefix`hook
- adds `product-picker-input` 
- adds `product-picker-page`
  - This is for testing purposes is ought to be removed later

I intentionally didn't generate the types to keep the PR small.
per-commit review is recommended.

Keeping this as draft, for early review on implementation and to test internally.

#### TODO

- ~documentation [reference](https://docs.commercetools.com/custom-applications/components/info-dialog)~ (separate PR)
- [x] Add tests for `product-picker-input`

##### When approved

- [x] remove `playground/product-picker-page`
- [x] add `changeset`
- [x] generate CTP types

### CODE REVIEW

File: packages/application-components/src/experimental/components/product-picker-input/product-picker-input.tsx
Comment (emmenko) [inline packages/application-components/src/experimental/components/product-picker-input/product-picker-input.tsx:196]: Be careful with this!

If the props change their reference on each re-render, you end up in an infinite loop here.

I bumped into a similar issue once: https://github.com/commercetools/merchant-center-application-kit/pull/1567

I suggest to exclude `onInitialLoadError` from the dependencies, as the other values are controlled by us.

File: packages/application-components/src/experimental/components/product-picker-input/product-picker-input.tsx
Comment (emmenko) [inline packages/application-components/src/experimental/components/product-picker-input/product-picker-input.tsx:205]: I'm not sure this component needs the horizontal constraint.

File: packages/application-components/src/experimental/components/product-picker-input/product-picker-input.tsx
Comment (emmenko) [inline packages/application-components/src/experimental/components/product-picker-input/product-picker-input.tsx:208]: Nit, in these cases I prefer to do

```ts
if (prefetchSelectedProductQuery.loading) {
  return ()
}

return ()
```

File: packages/application-components/src/experimental/components/shared/centered-loading-spinner/centered-loading-spinner.tsx
Comment (emmenko) [inline packages/application-components/src/experimental/components/shared/centered-loading-spinner/centered-loading-spinner.tsx:5]: Do we need `React.FC`?

File: packages/application-components/src/experimental/hooks/use-debounced-promise-callback/use-debounced-promise-callback.ts
Comment (emmenko) [inline packages/application-components/src/experimental/hooks/use-debounced-promise-callback/use-debounced-promise-callback.ts:7]: Shouldn't this be something like:

```suggestion
const useDebouncedPromiseCallback = <T>(
  promiseCallback: (...args: unknown[]) => Promise<T>,
  wait: number
) => useCallback(debounce(promiseCallback, wait), [promiseCallback]);
```

File: packages/application-components/src/experimental/hooks/use-load-with-prefix-search-fields/use-load-with-prefix-search-fields.ts
Comment (emmenko) [inline packages/application-components/src/experimental/hooks/use-load-with-prefix-search-fields/use-load-with-prefix-search-fields.ts:49]:
```suggestion
): string => {
```

File: packages/application-components/src/experimental/hooks/use-load-with-prefix-search-fields/use-load-with-prefix-search-fields.ts
Comment (emmenko) [inline packages/application-components/src/experimental/hooks/use-load-with-prefix-search-fields/use-load-with-prefix-search-fields.ts:60]: Nit: We should provide the return type, as we export this function.

File: packages/application-components/src/experimental/hooks/use-load-with-prefix-search-fields/use-load-with-prefix-search-fields.ts
Comment (emmenko) [inline packages/application-components/src/experimental/hooks/use-load-with-prefix-search-fields/use-load-with-prefix-search-fields.ts:64]:
```suggestion
    async ({
```

To make it clear that it returns a Promise

File: packages/application-components/src/utils/is-uuid.ts
Comment (emmenko) [inline packages/application-components/src/utils/is-uuid.ts:1]: Maybe we should use the `uuid.validate` function

https://www.npmjs.com/package/uuid#uuidvalidatestr

File: playground/src/routes.js
Comment (emmenko) [inline playground/src/routes.js:34]: We should remove this before merging.

**Review by emmenko — COMMENTED** [review]
I did a first review. Looks good overall, just some improvements here and there.

Great job so far! 💯

File: packages/application-components/src/experimental/components/product-picker-input/product-picker-input.tsx
Comment (adnasa) [reply packages/application-components/src/experimental/components/product-picker-input/product-picker-input.tsx:205]: I looked at this again. I think it makes sense to keep this for now.

File: packages/application-components/src/experimental/components/product-picker-input/product-picker-input.tsx
Comment (adnasa) [reply packages/application-components/src/experimental/components/product-picker-input/product-picker-input.tsx:75]: The thing about this, is that the generated types are in `application-shell` and they are not exported.
If we want to do this, we need to export the exports from `application-shell` then to be imported into `application-components`

I'm not really sure if we want that.

Comment (emmenko) [reply packages/application-components/src/experimental/components/product-picker-input/product-picker-input.tsx:75]: No need for that, you simply generate the types for the components package. Check the graphql config file.

We already do that for other packages as we can't really share those types.

File: packages/application-components/src/experimental/components/product-picker-input/product-picker-input.tsx
Comment (adnasa) [reply packages/application-components/src/experimental/components/product-picker-input/product-picker-input.tsx:1]: hmm

```
Duplicate declaration "React"
🎁 error @commercetools-frontend/application-components   1 | import type { ChangeEvent } from 'react';
🎁 error @commercetools-frontend/application-components > 2 | import React from 'react';
🎁 error @commercetools-frontend/application-components     |        ^^^^^
```

File: packages/application-components/src/experimental/components/product-picker-input/product-picker-input.tsx
Comment (adnasa) [inline packages/application-components/src/experimental/components/product-picker-input/product-picker-input.tsx:148]: now we skip the prefetch of product if the initial value was invalid.
this is preventive action to getting an `onInitialLoadError`

File: packages/application-components/src/experimental/hooks/use-debounced-promise-callback/use-debounced-promise-callback.ts
Comment (adnasa) [inline packages/application-components/src/experimental/hooks/use-debounced-promise-callback/use-debounced-promise-callback.ts:5]: I tried `unknown` but I got a pretty tricky TS error on the consumer side, which I spent too much time trying to resolve.
Saving this for another day

image

File: packages/application-components/src/experimental/components/product-picker-input/product-picker-input.spec.js
Comment (adnasa) [inline packages/application-components/src/experimental/components/product-picker-input/product-picker-input.spec.js:58]: I have a TODO to add `Product` into `test-data`.
it's about time

File: @types/commercetools-uikit__async-select-input/index.d.ts
Comment (adnasa) [inline @types/commercetools-uikit__async-select-input/index.d.ts:21]: `null` is the fallback value when consumer is able to convert a nullable `data` field into an option.
I prefer this over `undefined`. `null` derives directly from the Apollo result

**Comment by emmenko** [comment]
@adnasa FYI: I pushed a commit to improve a bunch of things.

File: packages/application-config/src/schema.ts
Comment (emmenko) [inline packages/application-config/src/schema.ts:24]: ??

**Review by emmenko — COMMENTED** [review]
I can take a look at the failing test

File: packages/application-components/src/experimental/components/internals/centered-loading-spinner/centered-loading-spinner.tsx
Comment (pa3) [inline packages/application-components/src/experimental/components/internals/centered-loading-spinner/centered-loading-spinner.tsx:7]: In the future (not sure worth doing now) the file can probably be moved outside of the `internals` scope and the `test-id` changed to be just `data-testid="centered-loading-spinner"`. I can foresee the need for this component in other places too.

File: packages/application-components/src/experimental/components/internals/option/option.tsx
Comment (pa3) [inline packages/application-components/src/experimental/components/internals/option/option.tsx:17]: Asking just out of personal curiosity: why not just 

```
const (props: TProps): JSX.Element => {
```

TS will complain that additional props are passed in this case?

File: packages/application-components/src/experimental/components/product-picker-input/product-picker-input.spec.tsx
Comment (pa3) [inline packages/application-components/src/experimental/components/product-picker-input/product-picker-input.spec.tsx:4]: Leftovers?

File: packages/application-components/src/experimental/hooks/use-load-with-prefix-search-fields/use-load-with-prefix-search-fields.spec.ts
Comment (pa3) [inline packages/application-components/src/experimental/hooks/use-load-with-prefix-search-fields/use-load-with-prefix-search-fields.spec.ts:9]: :heart_eyes:

**Review by pa3 — APPROVED** [review]
LGTM! :+1:

File: packages/application-components/src/experimental/components/product-picker-input/product-picker-input.spec.tsx
Comment (adnasa) [reply packages/application-components/src/experimental/components/product-picker-input/product-picker-input.spec.tsx:4]: yeah my bad. 
They are not supposed to be there :)

File: packages/application-components/src/experimental/components/internals/option/option.tsx
Comment (adnasa) [reply packages/application-components/src/experimental/components/internals/option/option.tsx:17]: hey, thanks for this question.
What I want to achieve here is that the consumer can pass more props than `props.data`.

as you can see in `L19`, we spread the props to the underlying component.
If the consumer implements a select option that consumes this, they can define their own _required_ props top of `TProps`.

At least that was my aim.

File: packages/application-components/src/experimental/components/internals/centered-loading-spinner/centered-loading-spinner.tsx
Comment (adnasa) [reply packages/application-components/src/experimental/components/internals/centered-loading-spinner/centered-loading-spinner.tsx:7]: yes I have always considered this, but I came to the same conclusion as you that this is premature

File: packages/application-components/src/experimental/components/internals/option/option.tsx
Comment (tdeekens) [reply packages/application-components/src/experimental/components/internals/option/option.tsx:17]: React.ReactNode and JSX.Element always confused me as to when to use which. So I'll just share that here too. A ReactNode is a ReactElement, a ReactFragment, a string, a number or an array of ReactNodes, or null, or undefined, or a boolean. A SX.Element is a ReactElement, with the generic type for props and type being any.

This always drove me nuts. I kept the rule of thumb that a ReactNode is what render on a class component returns and is used if you e.g. allow to return a string A JSX.Element is only for Functional Components.

That said: if an option can be a string or undefined then this might want to be ReactNode.

Comment (emmenko) [reply packages/application-components/src/experimental/components/internals/option/option.tsx:17]: A `JSX.Element` is what usually TS uses/infers when you don't explicitly define the return type. Which is also how  `FC` is declared.

image

So it's fine here. A `ReactNode` is what you use for example for typing `children`.

Comment (emmenko) [reply packages/application-components/src/experimental/components/internals/option/option.tsx:17]: > If the consumer implements a select option that consumes this, they can define their own required props top of TProps.

This component is only used internally, so we don't really need to make it unnecessarily generic.
You can keep it simple maybe.
I'm also fine either way.

Comment (tdeekens) [reply packages/application-components/src/experimental/components/internals/option/option.tsx:17]: What I said: JSX.Element is for React.FC. ReactNode for ClassComponent.

Comment (emmenko) [reply packages/application-components/src/experimental/components/internals/option/option.tsx:17]: You can always "force it" to return a `JSX.Element` by wrapping it into a fragment, for example:

```jsx
const Text = () => <>Hello</>
```

File: packages/application-components/src/experimental/components/product-picker-input/product-picker-input.spec.tsx
Comment (emmenko) [inline packages/application-components/src/experimental/components/product-picker-input/product-picker-input.spec.tsx:21]: Better maybe to consider the default props without a `value`. Then when you want to test with an existing value, you pass it explicitly.

File: packages/application-components/src/experimental/components/product-picker-input/product-picker-input.spec.tsx
Comment (emmenko) [inline packages/application-components/src/experimental/components/product-picker-input/product-picker-input.spec.tsx:84]: Note: as follow up we should maybe look into making test-utils more generic available. For instance, I don't like much that we use the appshell test-utils in this package.

File: packages/application-components/src/experimental/components/product-picker-input/product-picker-input.spec.tsx
Comment (emmenko) [inline packages/application-components/src/experimental/components/product-picker-input/product-picker-input.spec.tsx:107]: Do we also expect to show an error message/state here?

File: packages/application-components/src/experimental/components/product-picker-input/product-picker-input.spec.tsx
Comment (emmenko) [inline packages/application-components/src/experimental/components/product-picker-input/product-picker-input.spec.tsx:152]: I think this is not necessary, as it's implied by `findBy`

File: packages/application-components/src/experimental/components/product-picker-input/product-picker-input.spec.tsx
Comment (emmenko) [inline packages/application-components/src/experimental/components/product-picker-input/product-picker-input.spec.tsx:179]: Same

File: packages/application-components/src/experimental/components/product-picker-input/product-picker-input.spec.tsx
Comment (emmenko) [inline packages/application-components/src/experimental/components/product-picker-input/product-picker-input.spec.tsx:208]: Same

File: packages/application-components/src/experimental/components/product-picker-input/product-picker-input.spec.tsx
Comment (emmenko) [inline packages/application-components/src/experimental/components/product-picker-input/product-picker-input.spec.tsx:213]: `toHaveBeenCalledWith` ?

**Review by emmenko — COMMENTED** [review]
Looking good, just left some suggestions for the tests.

File: packages/application-components/src/experimental/components/product-picker-input/product-picker-input.spec.tsx
Comment (adnasa) [reply packages/application-components/src/experimental/components/product-picker-input/product-picker-input.spec.tsx:84]: yes I thought about that too. let's move this to an issue.

File: packages/application-components/src/experimental/components/product-picker-input/product-picker-input.spec.tsx
Comment (adnasa) [reply packages/application-components/src/experimental/components/product-picker-input/product-picker-input.spec.tsx:107]: with the status quo, I would not recommend that to display to the MC user.
The cases where the value is invalid;

1. user pastes in a invalid ID => do not send a request. displaying an error wouldn't make sense and if anything the consumer should decide that through `onInputChange`
2. the initial value was `null` <= this is a side effect of transformations

File: packages/application-config/src/schema.ts
Comment (emmenko) [reply packages/application-config/src/schema.ts:24]: Hmm this seems to be an artifact of the schema. Let's ignore this and keep it like this.

File: @types/commercetools-uikit__select-input/index.d.ts
Comment (adnasa) [inline @types/commercetools-uikit__select-input/index.d.ts:77]: 👍 
yeah I tried this too. works fine

**Review by emmenko — APPROVED** [review]
🚀

**Review by tdeekens — APPROVED** [review]

**Comment by adnasa** [comment]
Finally!! 🚀 
aaawwww, my baby going into `master` 😭 😭 😭  (tears of joy)

**Pull request merged**
